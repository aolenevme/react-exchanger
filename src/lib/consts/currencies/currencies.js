const currencies = Object.freeze({
    USD: "USD",
    EUR: "EUR",
    GBP: "GBP",
})

export default currencies;
