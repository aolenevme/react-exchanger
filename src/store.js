import {observable} from "mobx";

export default observable({
    time: "",
    timeColor: "#f88"
});
