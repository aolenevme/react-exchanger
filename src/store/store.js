import {observable} from "mobx";

import currencies from "../lib/consts/currencies/currencies.js";

export default observable({
    exchangeAmount: 12,
    rates: {
        target: 1.457,
        selected: 1.457
    },
    selectedCurrency: currencies.USD.abbreviation,
    targetCurrency: currencies.GBP.abbreviation,
    balances: {
        [currencies.USD.abbreviation]: 53.45,
        [currencies.EUR.abbreviation]: 100,
        [currencies.GBP.abbreviation]: 80.15
    }
});
