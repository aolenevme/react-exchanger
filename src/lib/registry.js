import store from "../store.js";

function updateIn(currentStore, payload) {
    const [path, value] = payload;

    path.reduce((acc, key, index) => {
        if (index === path.length - 1) {
            acc[key] = value;
        } else {
            return acc[key];
        }
    }, currentStore);
}

const storeEventsRegistry = {};

function dispatch(id, payload) {
    if (!id) {
        console.error("Enable to dispatch a store event without its id");

        return;
    }

    if (!storeEventsRegistry[id]) {
        console.error(
            `There isn\`t a registered handler for this store event id: ${id}`
        );

        return;
    }

    updateIn(store, storeEventsRegistry[id](store, payload));
}

function regEventStore(id, storeEvent) {
    if (!id) {
        console.error("Unable to register the event without its id");
    }

    if (!storeEvent) {
        console.error("Unable to register the store event without its handler");
    }

    if (storeEventsRegistry[id]) {
        console.error(
            `Unable to register the store event because it is already 
            registered: ${storeEventsRegistry[id]}`
        );
    }

    storeEventsRegistry[id] = storeEvent;
}

export {dispatch, regEventStore};
