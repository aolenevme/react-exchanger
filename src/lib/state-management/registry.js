import store from "../../store/store.js";

const storeEventsRegistry = {};

function dispatch(id, payload) {
    if (!id) {
        console.error("Enable to dispatch a store event without its id");

        return;
    }

    if (!storeEventsRegistry[id]) {
        console.error(
            `There isn\`t a registered handler for this store event id: ${id}`
        );

        return;
    }

    updateIn(store, storeEventsRegistry[id](store, payload));
}

function updateIn(currentStore, payload) {
    const [path, value] = payload;

    path.reduce((acc, key, index) => {
        if (index === path.length - 1) {
            acc[key] = value;
        } else {
            return acc[key];
        }
    }, currentStore);
}

function regEventStore(id, storeEvent) {
    if (!id) {
        console.error("Unable to register the event without its id");
    }

    if (!storeEvent) {
        console.error("Unable to register the store event without its handler");
    }

    if (storeEventsRegistry[id]) {
        console.error(
            `Unable to register the store event because it is already 
            registered: ${storeEventsRegistry[id]}`
        );
    }

    storeEventsRegistry[id] = storeEvent;
}

export {dispatch, regEventStore};
