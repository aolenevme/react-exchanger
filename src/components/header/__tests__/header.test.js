import React from "react";
import renderer from "react-test-renderer";

import Header from "../header.js";
import * as registry from "../../../lib/state-management/registry.js";
import * as store from "../../../store/store.js";
import currencies from "../../../lib/consts/currencies/currencies.js";
import MUTATE_STORE from "../../../events/mutate-store.js";
import checkStoreMutations from "../../../lib/tests/check-store-mutations.js";

describe("<Header />", () => {
    beforeEach(() => {
        registry.dispatch = jest.fn();
    });

    afterEach(() => {
        registry.dispatch.mockClear();
    });

    describe("rendering", () => {
        it("is rendered correctly by default", () => {
            store.default = {
                exchangeAmount: 123,
                selectedCurrency: currencies.USD.abbreviation,
                targetCurrency: currencies.EUR.abbreviation
            };

            expect(renderer
                .create(<Header />)
                .toJSON()).toMatchSnapshot();
        });
    });

    describe("disabled rendering", () => {
        it("is rendering correctly when currencies are equal", () => {
            store.default = {
                selectedCurrency: currencies.USD.abbreviation,
                targetCurrency: currencies.USD.abbreviation,
                exchangeAmount: 123
            };

            expect(renderer
                .create(<Header />)
                .toJSON()).toMatchSnapshot();
        });

        it("is rendering correctly when no exchange amount is typed", () => {
            store.default = {
                selectedCurrency: currencies.USD.abbreviation,
                targetCurrency: currencies.EUR.abbreviation,
                exchangeAmount: ""
            };

            expect(renderer
                .create(<Header />)
                .toJSON()).toMatchSnapshot();
        });
    });

    describe("events", () => {
        it("resets value on Cancel", () => {
            renderer
                .create(<Header />)
                .toJSON().children[0].props.onClick();

            expect(registry.dispatch).toHaveBeenCalled();
        });

        it("transfer money on Exchange and resets exchange amount", () => {
            store.default = {
                exchangeAmount: 10,
                selectedCurrency: currencies.USD.abbreviation,
                targetCurrency: currencies.EUR.abbreviation,
                rates: {
                    target: 0.9
                },
                balances: {
                    [currencies.USD.abbreviation]: 50,
                    [currencies.EUR.abbreviation]: 40
                }
            };

            renderer
                .create(<Header />)
                // eslint-disable-next-line no-magic-numbers
                .toJSON().children[2].props.onClick();

            checkStoreMutations(0, MUTATE_STORE, [["balances", currencies.USD.abbreviation], "40.00"]);
            checkStoreMutations(1, MUTATE_STORE, [["balances", currencies.EUR.abbreviation], "49.00"]);
            // eslint-disable-next-line no-magic-numbers
            checkStoreMutations(2, MUTATE_STORE, [["exchangeAmount"], ""]);
        });
    });
});
