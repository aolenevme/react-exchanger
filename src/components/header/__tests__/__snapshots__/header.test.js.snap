// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Header /> is rendered correctly 1`] = `
.c1 {
  border-color: transparent;
  background-color: transparent;
  color: #f3f6ff;
  -webkit-tap-highlight-color: transparent;
  font-size: 1rem;
  cursor: pointer;
  -webkit-transition: color 0.1s ease 0s;
  transition: color 0.1s ease 0s;
}

.c1:active {
  color: #5591f7;
}

.c1:disabled {
  cursor: unset;
}

.c1:hover {
  color: #5591f7;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding: 0.2rem 0.5rem;
  min-height: 2rem;
  border: 1px solid #5591f7;
  border-radius: 4px;
  background-color: #2c5bcd;
  color: #f3f6ff;
  font-weight: 400;
  cursor: pointer;
  -webkit-transition: border-color 0.1s ease 0s;
  transition: border-color 0.1s ease 0s;
}

.c2:active {
  border-color: #f3f6ff;
}

.c2:hover {
  border-color: #f3f6ff;
}

.c3 {
  width: 100%;
  text-align: center;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  padding: 0 1rem;
  height: 2rem;
}

<div
  className="c0"
>
  <button
    className="c1"
    disabled={false}
    onClick={[Function]}
    type="button"
  >
    Cancel
  </button>
  <div
    className="c2"
  >
    <span
      className="c3"
    >
      $1 = Â£1.4570
    </span>
  </div>
  <button
    className="c1"
    disabled={false}
    onClick={[Function]}
    type="button"
  >
    Exchange
  </button>
</div>
`;
